File: modules/tests/src/commonTest/kotlin/primitives/PingPrimitive.kt
package primitives

import natives.test.MyDictionary
import natives.test.MyEnum
import natives.test.pingBoolean
import natives.test.pingByte
import natives.test.pingChar
import natives.test.pingDouble
import natives.test.pingFloat
import natives.test.pingInt
import natives.test.pingLong
import natives.test.pingString
import natives.test.pingShort
import withLib
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class PingPrimitive {

    @Test
    fun pingChar() = withLib {
        assertEquals('a', pingChar('a'))
    }

    @Test
    fun pingBoolean() = withLib {
        assertEquals(true, pingBoolean(true))
    }

    @Test
    fun pingByte() = withLib {
        assertEquals(99.toByte(), pingByte(99.toByte()))
    }

    @Test
    fun pingShort() = withLib {
        assertEquals(99.toShort(), pingShort(99.toShort()))
    }

    @Test
    fun pingInt() = withLib {
        assertEquals(99, pingInt(99))
    }

    @Test
    fun pingLong() = withLib {
        assertEquals(9223372036854775805L, pingLong(9223372036854775805L))
    }

    @Test
    fun pingFloat() = withLib {
        assertEquals(99f, pingFloat(99f))
    }

    @Test
    fun pingDouble() = withLib {
        assertEquals(99.0, pingDouble(99.0))
    }

    @Test
    fun pingString() = withLib {
        assertEquals("test string", pingString("test string"))
    }

    @Test
    fun pingStringN() = withLib {
        assertNull(natives.test.pingStringN(null))
    }

    @Test
    fun pingEnum() = withLib {
        assertEquals(MyEnum.CASE2, natives.test.pingEnum(MyEnum.CASE2))
    }

    @Test
    fun pingDictionary() = withLib {
        assertEquals(
            MyDictionary(1, 2, 3, 4),
            natives.test.pingDictionary(MyDictionary(1, 2, 3, 4))
        )
    }

    @Test
    fun pingDictionaryN() = withLib {
        assertNull(natives.test.pingDictionaryN(null))
    }

}

File: modules/tests/natives/test/src/primitives/ping_primitive.c
#include <api.h>

uint16_t ping_char(const uint16_t arg) {
    return arg;
}

bool ping_boolean(const bool arg) {
    return arg;
}

int8_t ping_byte(const int8_t arg) {
    return arg;
}

uint8_t ping_ubyte(const uint8_t arg) {
    return arg;
}

int16_t ping_short(const int16_t arg) {
    return arg;
}

uint16_t ping_ushort(const uint16_t arg) {
    return arg;
}

int32_t ping_int(const int32_t arg) {
    return arg;
}

uint32_t ping_uint(const uint32_t arg) {
    return arg;
}

int64_t ping_long(const int64_t arg) {
    return arg;
}

uint64_t ping_ulong(const uint64_t arg) {
    return arg;
}

float ping_float(const float arg) {
    return arg;
}

double ping_double(const double arg) {
    return arg;
}

KString* ping_string(KString* arg) {
    return arg->clone(arg);
}

KString* ping_string_n(KString* arg) {
    return arg;
}

MyEnum ping_enum(const MyEnum arg) {
    return arg;
}

MyDictionary* ping_dictionary(MyDictionary* arg) {
    return arg->clone(arg);
}

MyDictionary* ping_dictionary_n(MyDictionary* arg) {
    return arg;
}

RC_MyInterface* ping_interface(RC_MyInterface* arg) {
    return arg->clone(arg);
}

RC_MyInterface* ping_interface_n(RC_MyInterface* arg) {
    return arg;
}